flaml/autogen/agent/responsive_agent.py

```diff
@@ -260,11 +261,19 @@ def generate_reply(self, messages: List[Dict], default_reply: Union[str, Dict] =
             return func_return
         if self._code_execution_config is False:
             return default_reply if self.oai_config is False else self._oai_reply(messages)
-        code_blocks = extract_code(message["content"])
+        
+        # search from the most recent message to the oldest message for code blocks
+        for msg in reversed(messages):
+            code_blocks = extract_code(msg["content"])
+            print("code_blocks", code_blocks)
+            if len(code_blocks) == 1 and code_blocks[0][0] != UNKNOWN:
+                break
```

Comment:
This is problematic because (1) the same code will get re-executed if no new code is suggested. (2) if human prevents code execution in one turn, the code may get executed in later turn.

Reply:
I see. However, simply checking the most recent message might also be problematic. For example, when in the `code writer` -> `code-reviewer` -> `code-runner` pattern.

To prevent 1), maybe we can simply check the "new messages" for a specific agent? AKA the messages after that specific agent's most recent message. To resolve (2), we might need another llm call to retrieve the running code from chatting history?

Reply:
I pushed a change to use llm to find code blocks when regex fails in #1154. That could be an alternative to the approach you use now.